prometheus/client.go

```diff
@@ -38,18 +42,62 @@ type Client struct {
 	api prom_v1.API
 }
 
+type tokenRoundTripper struct {
+	bearerToken string
+	originalRT  http.RoundTripper
+}
+
+func (tokenRT *tokenRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", "Bearer "+tokenRT.bearerToken)
+	return tokenRT.originalRT.RoundTrip(req)
+}
+
+func newTokenRoundTripper(token string, rt http.RoundTripper) http.RoundTripper {
+	return &tokenRoundTripper{token, rt}
+}
+
 // NewClient creates a new client to the Prometheus API.
 // It returns an error on any problem.
 func NewClient() (*Client, error) {
 	cfg := config.Get().ExternalServices.Prometheus
 	clientConfig := api.Config{Address: cfg.URL}
+
+	transportConfig := api.DefaultRoundTripper.(*http.Transport)
+
 	if cfg.InsecureSkipVerify {
-		transportConfig := api.DefaultRoundTripper.(*http.Transport)
 		transportConfig.TLSClientConfig = &tls.Config{
 			InsecureSkipVerify: true,
 		}
+	}
+
+	if cfg.CAFile != "" {
+		certPool := x509.NewCertPool()
+		cert, err := ioutil.ReadFile(cfg.CAFile)
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Prometheus root CA certificates: %s", err)
+		}
+
+		certPool.AppendCertsFromPEM(cert)
+
+		transportConfig.TLSClientConfig = &tls.Config{
+			RootCAs: certPool,
+		}
+	}
+
+	if cfg.Auth == "bearer" {
+		token, err := kubernetes.GetKialiToken()
```

Comment:
Here you're reading the Kiali SA token to pass it to prometheus ... i don't think it's what we want, or is it? I would imagine we had to "forward" the user token that comes from the request that triggers the prom api call.

Reply:
We want to use the Kiali service account token here and not the user token.

The user token is still used, but its used by Kiali determine if the user has access to those metrics or not. Its Kiali that does the metric filtering, not Prometheus (since Prometheus doesn't have the concept of metric permissions).

For the other secured components, such as Grafana and Jaeger, we do want to use the user token. Its only for Prometheus that we don't since Kiali already does the filtering.

Reply:
Why not comment the code with this kind of information?  No one will remember all the good info in the PR.

Reply:
+1
To add extended comments in the code for this.
It will help a lot to maintain it for others.

Thanks !